music21/musicxml/m21ToXml.py

```diff
@@ -1735,6 +1736,66 @@ def parseFlatScore(self):
         pp.parse()
         self.partExporterList.append(pp)
 
+    def prePartProcess(self):
+        '''
+        Scans the parts/measures/voices in each joinable group to compute
+        the voiceNumberOffset needed in each part.  Parts that aren't in a
+        joinable group get offset 0.
+
+        Called automatically by .parse()
+        '''
+        partExporterForPart: t.Dict[stream.Part, PartExporter] = {}
+
+        # Compute partExp.staffGroup and partExp.maxNumVoices
+        for partExp in self.partExporterList:
+            partExporterForPart[partExp.stream] = partExp
+
+            joinableGroup = None
+            for sg in self.groupsToJoin:
+                if partExp.stream in sg:
+                    joinableGroup = sg
+                    break
+
+            partExp.staffGroup = joinableGroup
+            if partExp.staffGroup is None:
+                partExp.maxVoiceNum = 0
+                continue
+
+            # part is in a staffGroup, so we need to find the maximum number
+            # of voices used in any measure in this part.  We also find the
+            # maximum "low enough to not be considered a memory location"
+            # voice id.  The maxVoiceNum for a part is the higher of
+            # maxNumVoices-1 and maxNonMemLocationVoiceId.
+            maxNumVoices = 0
+            maxNonMemLocationVoiceId = 0
+            for measure in partExp.stream[stream.Measure]:
+                voices = tuple(measure[stream.Voice])
+                maxNumVoices = max(maxNumVoices, len(voices))
+                for voice in voices:
+                    if (isinstance(voice.id, int)
+                            and voice.id < defaults.minIdNumberToConsiderMemoryLocation):
+                        maxNonMemLocationVoiceId = max(maxNonMemLocationVoiceId, voice.id)
+            partExp.maxVoiceNum = max(maxNumVoices - 1, maxNonMemLocationVoiceId)
```

Comment:
Could you explain why the "minus 1" comes in? If there are two voices, shouldn't the maxVoiceNum be 2? (A 1 gets added back, below, which kind of adds to the understanding barrier for me.)

Reply:
Yeah, I was thinking in 0-based indices (as one does), and then converting to 1-based at the last minute, but you're right, if I keep voice numbers 1-based throughout (as they are in MusicXML) it's much clearer.